lib/plugins/leetcode.js

```diff
@@ -594,27 +620,50 @@ plugin.githubLogin = function(user, cb) {
       if (resp.statusCode !== 200) {
         return cb('GitHub login failed');
       }
-      _request.get({url: leetcodeUrl}, function(e, resp, body) {
-        const redirectUri = resp.request.uri.href;
-        if (redirectUri !== 'https://leetcode.com/') {
-          return cb('GitHub login failed or GitHub did not link to LeetCode');
+      if (resp.request.uri.href === urls.github_tf_redirect) {
+        cb('Your GitHub are using two-factor authentication');
+        // read two-factor code must be sync.
+        const twoFactorcode = require('prompt-sync')()('Please enter your two-factor code: ');
+        const authenticityTokenTwoFactor = body.match(/name="authenticity_token" value="(.*?)"/);
+        if (authenticityTokenTwoFactor === null) {
+          return cb('Get GitHub two-factor token failed');
         }
-        const cookieData = parseCookie(resp.request.headers.cookie, cb);
-        saveAndGetUser(user, cb, cookieData);
-      });
+        const optionsTwoFactor = {
+          url:     urls.github_tf_session_request,
+          method:  'POST',
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded',
+          },
+          followAllRedirects: true,
+          form:               {
+            'otp':                twoFactorcode,
+            'authenticity_token': authenticityTokenTwoFactor[1],
+            'utf8':               encodeURIComponent('✓'),
+          },
+        };
+        _request(optionsTwoFactor, function(e, resp, body) {
+          if (resp.request.uri.href === urls.github_tf_session_request) {
+            return cb('Invalid two-factor code please check');
+          }
+          requestLeetcodeAndSave(_request, leetcodeUrl, user, cb, config.sys.login_methods.GitHub);
+        });
+      } else {
+        requestLeetcodeAndSave(_request, leetcodeUrl, user, cb, config.sys.login_methods.GitHub);
```

Comment:
Can the calling of `requestLeetcodeAndSave` in line 648 and line 651 merged into one call -- move it out of the if-else block?

Reply:
Sorry, these two are with different session(_request), the first one is normal, the second one is two-factor. Both need.

Reply:
I see. How about

```js
if (resp.request.uri.href !== urls.github_tf_redirect) {
  return requestLeetcodeAndSave(_request, leetcodeUrl, user, cb, config.sys.login_methods.GitHub);
}

cb('Your GitHub are using two-factor authentication');
// read two-factor code must be sync.
const twoFactorcode = require('prompt-sync')()('Please enter your two-factor code: ');
const authenticityTokenTwoFactor = body.match(/name="authenticity_token" value="(.*?)"/);
if (authenticityTokenTwoFactor === null) {
  return cb('Get GitHub two-factor token failed');
}
const optionsTwoFactor = {
  url:     urls.github_tf_session_request,
  method:  'POST',
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded',
  },
  followAllRedirects: true,
  form:               {
    'otp':                twoFactorcode,
    'authenticity_token': authenticityTokenTwoFactor[1],
    'utf8':               encodeURIComponent('✓'),
  },
};
_request(optionsTwoFactor, function(e, resp, body) {
  if (resp.request.uri.href === urls.github_tf_session_request) {
    return cb('Invalid two-factor code please check');
  }
  requestLeetcodeAndSave(_request, leetcodeUrl, user, cb, config.sys.login_methods.GitHub);
});
```

which can save some indentions.

Reply:
I

Reply:
Yes! Very clean code. Thank you.